translation_layer/jpl-capi.js

```diff
@@ -0,0 +1,150 @@
+/*globals console, window*/
+define ('main',['require','jquery','ExtendedModel','api/snapshot/adapters/BackboneAdapter','sim-common/Detect','sim-common/VersionCheck','text!sim-common/templates/notSupported.html','api/snapshot/adapters/BackboneAdapter','api/snapshot/Controller','api/snapshot/Transporter','api/snapshot/SimCapiValue'],function(require){
+
+    var $ = require('jquery');
+    var $body;
+
+    var ExtendedModel = require('ExtendedModel');
+    var adapter = require('api/snapshot/adapters/BackboneAdapter').getInstance();
+
+    var Capi = ExtendedModel.extend({});
+    var capi;
+
+    require('sim-common/Detect');
+    require('sim-common/VersionCheck');
+
+    var notSupportedTemplate = require('text!sim-common/templates/notSupported.html');
+    var BackboneAdapter = require ('api/snapshot/adapters/BackboneAdapter').getInstance();
+    var Controller = require('api/snapshot/Controller');
+    var Transporter = require('api/snapshot/Transporter').getInstance();
+    var Types = require('api/snapshot/SimCapiValue').TYPES;
+
+	var initialized = false;
+
+    window.receiveExposeFromRanger = function (name, type, value, allowedValues)
+    {
+        if (!initialized)
+        {
+            Controller.notifyOnReady();
+            capi = new Capi();
+            capi.on('change', onCapiChange);
+            initialized = true;
+        }
+
+        var typeConverter = generateTypeConverter(type);
+        value = typeConverter(value);
+
+        var descObj = {};
+
+        //console.log("allowed values: " + allowedValues);
+
+        if (allowedValues)
+        {
+            descObj.allowedValues = allowedValues.map(typeConverter);
+        }
+
+        capi.set(name, value);
+        adapter.expose(name, capi, descObj);
+    };
+
+	    window.receiveValueFromRanger = function (name, type, value) {
+        if (initialized) {
+            value = generateTypeConverter(type)(value);
+            capi.set(name, value);
+        }
```

Comment:
Right - half of this code was written in anticipation of getting the Ranger code from JPL, but the end deliverable we received didn't give us full access. receiveValueFromRanger was stubbed to be the function used on their end to set CAPI, but you are correct in observing we don't get to use it. At best, we can set up an acknowledgement function that *might* be capable of receiving data that can be set into the CAPI. In practice, our use of CAPI in this particular sim is one way - delivering messages to Ranger as their acknowledgement system doesn't always respond with the data in the same format it's sent in. Ornery.

receiveExposeFromRanger() is overloaded. Since I can't trigger commands and calls directly from Ranger, I play it safe by checking for initialization before exposing anything. There is API from Ranger to confirm that it has started, but I haven't wrangled this to work with their acknowledgement system yet.